src/PostgREST/Request/QueryParams.hs

```diff
@@ -0,0 +1,489 @@
+    offsets = first (replaceLast "limit") <$> filter (endingIn ["offset"] . fst) nonemptyParams
+    lookupParam :: Text -> Maybe Text
+    lookupParam needle = toS <$> join (L.lookup needle qParams)
+    nonemptyParams = mapMaybe (\(k, v) -> (k,) <$> v) qParams
+
+    qString = HTTP.parseQueryReplacePlus True qs
+
+    qParams = [(T.decodeUtf8 k, T.decodeUtf8 <$> v)|(k,v) <- qString]
+
+    canonical =
+      BS.pack $ HTTP.urlEncodeVars
+        . L.sortOn fst
+        . map (join (***) BS.unpack . second (fromMaybe mempty))
+        $ qString
+
+    endingIn:: [Text] -> Text -> Bool
+    endingIn xx key = lastWord `elem` xx
+      where lastWord = L.last $ T.split (== '.') key
+
+    (filters, params) = L.partition isParam filtersAndParams
+    isParam (k, v) = isEmbedPath k || hasOperator v || hasFtsOperator v
+
+    filtersAndParams = filter (isFilterOrParam . fst)  nonemptyParams
+    isFilterOrParam k = not (endingIn reservedEmbeddable k) && notElem k reserved
+    reserved = ["select", "columns", "on_conflict"]
+    reservedEmbeddable = ["order", "limit", "offset", "and", "or"]
+
+    (filtersNotRoot, filtersRoot) = L.partition isNotRoot filters
+    isNotRoot = flip T.isInfixOf "." . fst
+
+    -- TODO: These checks are redundant to the parsers, should use parsers to differentiate params
+    hasOperator val =
+      case T.splitOn "." val of
+        "not" : _ : _ -> True
+        "is" : _      -> True
+        "in" : _      -> True
+        x : _         -> isJust (operator x) || isJust (ftsOperator x)
+        _             -> False
+
+    hasFtsOperator val =
+      case T.splitOn "(" val of
+        x : _ : _ -> isJust $ ftsOperator x
+        _         -> False
+
+    isEmbedPath = T.isInfixOf "."
+    replaceLast x s = T.intercalate "." $ L.init (T.split (=='.') s) <> [x]
+
+    ranges :: M.HashMap Text (Range Integer)
+    ranges = M.unionWith f limitParams offsetParams
+      where
+        f rl ro = Range (BoundaryBelow o) (BoundaryAbove $ o + l - 1)
+          where
+            l = fromMaybe 0 $ rangeLimit rl
+            o = rangeOffset ro
+
+        limitParams =
+          M.fromList [(k, restrictRange (readMaybe $ toS v) allRange) | (k,v) <- limits]
+
+        offsetParams =
+          M.fromList [(k, maybe allRange rangeGeq (readMaybe $ toS v)) | (k,v) <- offsets]
+
+operator :: Text -> Maybe SingleValOperator
+operator = \case
+  "eq"    -> Just OpEqual
+  "gte"   -> Just OpGreaterThanEqual
+  "gt"    -> Just OpGreaterThan
+  "lte"   -> Just OpLessThanEqual
+  "lt"    -> Just OpLessThan
+  "neq"   -> Just OpNotEqual
+  "like"  -> Just OpLike
+  "ilike" -> Just OpILike
+  "cs"    -> Just OpContains
+  "cd"    -> Just OpContained
+  "ov"    -> Just OpOverlap
+  "sl"    -> Just OpStrictlyLeft
+  "sr"    -> Just OpStrictlyRight
+  "nxr"   -> Just OpNotExtendsRight
+  "nxl"   -> Just OpNotExtendsLeft
+  "adj"   -> Just OpAdjacent
+  _       -> Nothing
+
+ftsOperator :: Text -> Maybe FtsOperator
```

Comment:
I think the naming of `SingleValOperator` vs `FtsOperator` could be improved. They both really only have a single value, but one has arguments, the other does not, right?

What about:
- `Operator` vs `OperatorWithArg` or
- `OperatorWithoutArg` vs `OperatorWithArg` or
- `SimpleOperator` vs `OperatorWithArg`

This naming thing probably extends to the parser, too, because it's not really relevant that those operators are "fts" or not.

Reply:
Ah, but it does matter: The full text search operators have an optional language parameter, the single val operators do not. I followed the naming of the existing parsers here, would not change it in this PR.

Reply:
Mh. I think I didn't make my point clear enough. Maybe this code snippet tells the story a bit better:

```haskell
data Operation
  = Op SingleValOperator SingleVal
  | In ListVal
  | Is TrileanVal
  | Fts FtsOperator (Maybe Language) SingleVal
```

Both `SingleValOperator` and `FtsOperator` take a `SingleVal`. That means the `SingleVal` part in `SingleValOperator` is not really describing it well.

I agree with `FtsOperator` - that name was there before already. But `SingleValOperator` was not. It was called `Operator` before. Maybe just rename it to `SimpleOperator` or so - I think that's what you meant to say.